ift6759_train_Glow_2mTemp.py

```diff
@@ -0,0 +1,144 @@
+    """
+    "Spatial" MAE auxiliary loss for generator. Penalizes outputs
+    which violate spatial average preservation between input and output.
+    """
+    kernel = tf.ones((scale,scale,1,1)) / (scale**2.)
+    def _spatial_mse(x_in, y_pred):
+        x_avg = tf.nn.conv2d(x_in, kernel, strides=(stride, stride), padding='VALID')
+        y_avg = tf.nn.conv2d(y_pred, kernel, strides=(stride, stride), padding='VALID')
+        return tf.math.reduce_mean(tf.math.abs(y_avg - x_avg))
+    return _spatial_mse
+
+
+# load the data
+zarr_lr = xr.open_zarr('data/processed/temp/5625/temp_5625_processed.zarr')
+zarr_hr = xr.open_zarr('data/processed/temp/1406/temp_1406_processed.zarr')
+
+# center it to zero
+zarr_lr, monthly_means_lr = remove_monthly_means(zarr_lr, time_dim='date')
+zarr_hr, monthly_means_hr = remove_monthly_means(zarr_hr, time_dim='date')
+
+# make data numpy arrays (unsuited for large data):
+# each ndarray have shape [date, lat, lon]
+ndarray_lr = zarr_lr.to_array().to_numpy().squeeze()
+ndarray_hr = zarr_hr.to_array().to_numpy().squeeze()
+
+# Technically, we shouldn't have to do this, but we need to adjust our
+# preprocessing so the images are square with sizes of the powers of 2
+# meaning (16,16) or (32,32) or (64,64), etc..
+ndarray_lr = ndarray_lr
+ndarray_hr = ndarray_hr
+
+# defining tensorflow datasets
+# this batches the data along the date axis to yield
+# 1096 samples of shape [lat, lon]
+# We keep the last eighth of the sample for test set.
+assert len(ndarray_lr)==len(ndarray_hr)
+test_size = int(1/8 * len(ndarray_lr))
+train_size = len(ndarray_lr) - test_size
+
+dataset_train_lr = tf.data.Dataset.from_tensor_slices(ndarray_lr[:-test_size])
+dataset_train_hr = tf.data.Dataset.from_tensor_slices(ndarray_hr[:-test_size])
+dataset_test_lr = tf.data.Dataset.from_tensor_slices(ndarray_lr[-test_size:])
+dataset_test_hr = tf.data.Dataset.from_tensor_slices(ndarray_hr[-test_size:])
+
+# We need to naively upsample the low res data so it has the same
+# dimensionality as the high res using the nearest neighbour algo
+# We first add a channel to all images
+dataset_train_lr = dataset_train_lr.map(lambda x: x[:,:,None])
+dataset_test_lr = dataset_test_lr.map(lambda x: x[:,:,None])
+dataset_train_hr = dataset_train_hr.map(lambda x: x[:,:,None])
+dataset_test_hr = dataset_test_hr.map(lambda x: x[:,:,None])
+# Then upsample the low res datasets
+lat_hr, lon_hr = ndarray_hr.shape[1:]
+dataset_train_lr = dataset_train_lr.map(upsample(lat_hr, lon_hr, tf.image.ResizeMethod.NEAREST_NEIGHBOR))
+dataset_test_lr = dataset_test_lr.map(upsample(lat_hr, lon_hr, tf.image.ResizeMethod.NEAREST_NEIGHBOR))
+
+# zipping the data together and shuffling each dataset individually
+# for "unsupervised learning"
+train_ds = preprocess_vds(dataset_train_lr, dataset_train_hr, batch_size=10, buffer_size=train_size, supervised=False)
+test_ds = preprocess_vds(dataset_test_lr, dataset_test_hr, batch_size=10, buffer_size=test_size, supervised=False)
```

Comment:
is this just a simple train/test split? do we not want a third test set?

Reply:
Ohh I did as they do in groenke, but you're right. This would be best practice!

Reply:
on second thought, if we're just training one model (i.e. no hyperparameter optimization), i guess it doesn't matter

Reply:
That's right